src/library/filescanner.c

```diff
@@ -1816,12 +1816,62 @@ playlist_add_path(FILE *fp, int pl_id, const char *path)
   return 0;
 }
 
+static int
+ensure_file_record(const char *path)
+{
+  struct query_params qp;
+  struct db_media_file_info dbmfi;
+  struct media_file_info mfi;
+  bool found;
+  int ret;
+
+  memset(&qp, 0, sizeof(struct query_params));
+  qp.type = Q_ITEMS;
+  qp.sort = S_ARTIST;
+  qp.idx_type = I_NONE;
+  qp.filter = db_mprintf("(f.path = %Q)", path);
+
+  ret = db_query_start(&qp);
+  if (ret < 0)
+    {
+      db_query_end(&qp);
+      free(qp.filter);
+      return -1;
+    }
+
+  found = ((ret = db_query_fetch_file(&qp, &dbmfi)) == 0) && (dbmfi.id);
+  db_query_end(&qp);
+  free(qp.filter);
+
+  while (!found)
```

Comment:
What is the benefit of the while-loop?

Reply:
Generally: avoid `goto`, avoid deep nesting of `if`

Here, the second inner part could also live in an `else` branch.

However, sInce it is not always clear in advance, how many error conditions will arise, I just use this as default pattern to see the `good` case flow on the same indentation level (`if` sections == error handling).

In the final version, this idiom can be converted info a nested `if`/`else` for the simple cases.